Theme: split classes and exceptions in namespace summary

vlastavesely opened the pull request:
In older ApiGen, there were classes and exceptions summaries separated.

## Before:
![before](https://user-images.githubusercontent.com/22521379/27759704-7040a274-5e37-11e7-85f7-3be678a8dcdf.png)

## After
![after](https://user-images.githubusercontent.com/22521379/27759706-7a692582-5e37-11e7-9cc7-64bb2916727d.png)

TomasVotruba commented on src/Generator/NamespaceGenerator.php:
```diff
@@ -45,6 +45,16 @@ private function generateForNamespace(
+        $classes = $exceptions = [];
+
+        foreach ($namespaceReflectionCollector->getClassReflections($namespace) as $class) {
+            if (strtolower(substr($class->getName(), -9)) !== 'exception') {
```
Here can be used `is_a` function to check if the class is exception:

```php
is_a($class, Exception::class, true)
```

Name is less reliable than this.

vlastavesely replied:
I tried it but it does not seem to work. What about `$classReflection->implementsInterface('Throwable')`? That covers not only classes derived from `Exception` but also any other throwable classes that may not inherit from the `Exception`.

TomasVotruba commented:
I like this!

Just one detail and it's ready

TomasVotruba commented on packages/ThemeDefault/src/@elementlist.latte:
```diff
@@ -36,3 +36,10 @@
     </table>
 {/if}
+
+{if $exceptions}
```
Are `EmptyNamespaceGenerator` and `NamespaceGenerator` only places where this template `@elementlist.latte` is used?

I think it's also in [IndexGenerator template](https://github.com/ApiGen/ApiGen/blob/682fc8f43c463884bfef5fd0c8d56c403c8a5608/packages/ThemeDefault/src/index.latte#L20)

Maybe it would make sense to distinquish those already on `ReflectionStorage` level: https://github.com/ApiGen/ApiGen/blob/682fc8f43c463884bfef5fd0c8d56c403c8a5608/packages/Reflection/src/ReflectionStorage.php#L81

vlastavesely commented:
Oh yes, moving into `ReflectionStorage` makes sense. I'll do it.

vlastavesely commented:
What do you prefer on `classes.html` page - list of classes and list of exceptions under it or new page `exceptions.html`?

TomasVotruba commented:
Standalone exception page would make more sense, as it would be 1:1 to `ReflectionStorage`.

vlastavesely commented:
* added exceptions summary in `exceptions.html`
* link to the summary added into header (after "Overview", "Classes"...)
* added `ExceptionGenerator` - it uses `class.latte` (for exception is class actually)
* `exceptions.html` added into StepCountter but still, there is 1 item missing in the counter :(
* `ExceptionsGenerator` and `ExceptionGenerator` covered by tests

TomasVotruba commented on packages/Reflection/src/ReflectionStorage.php:
```diff
@@ -46,7 +59,11 @@ public function addClassReflections(array $classReflections): void
         sort($classReflections);
         foreach ($classReflections as $classReflection) {
-            $this->classReflections[$classReflection->getName()] = $classReflection;
+            if ($classReflection->implementsInterface('Throwable')) {
```
There could be used class instead of string: `Throwable::class`, it's easier to refactor and more future proof

vlastavesely replied:
ok

TomasVotruba commented on packages/ThemeDefault/src/partial/menu.latte:
```diff
@@ -34,6 +34,12 @@
+                            <li>
+                                <a n:class="$activePage === exceptions ? active" href="exceptions.html">Exceptions</a>
+                            </li>
+                        {/if}
```
Very nice!

TomasVotruba commented on src/EventSubscriber/ElementsTemplateVariablesEventSubscriber.php:
```diff
@@ -53,6 +53,7 @@ public function loadTemplateVariables(CreateTemplateEvent $createTemplateEvent):
         $parameterBag->addParameters([
             'allNamespaces' => $this->namespaceReflectionCollector->getNamespaces(),
             'allClasses' => $this->reflectionStorage->getClassReflections(),
+            'allExceptions' => $this->reflectionStorage->getExceptionReflections(),
```
That's the right place, I'm glad it's obvious.
:+1:

TomasVotruba commented on src/Console/Progress/StepCounter.php:
```diff
@@ -51,6 +53,10 @@ private function getOverviewPagesCount(): int
         }
 
+        if (count($this->reflectionStorage->getExceptionReflections())) {
+            $count++; // exceptions.html
```
:+1:  for comment

TomasVotruba commented on src/Generator/ExceptionGenerator.php:
```diff
@@ -0,0 +1,82 @@
+    {
+        $this->templateRenderer->renderToFile(
+            $this->configuration->getTemplateByName('class'),
+            $this->configuration->getDestinationWithPrefixName('class-', $exceptionReflection->getName()),
```
I would probably prefer `exception-` here, to be 1:1 with ReflectionStorage.

It was and might be confusing in the future, if something is class x exception and something is only class.

vlastavesely replied:
ok, working on it.

TomasVotruba replied:
Thank you

TomasVotruba commented on tests/Generator/ExceptionGeneratorTest.php:
```diff
@@ -0,0 +1,36 @@
+            TEMP_DIR . '/source-class-ApiGen.Tests.Generator.Source.SomeException.html'
+        );
+    }
+}
```
Beautiful test :clap:

TomasVotruba commented:
> exceptions.html added into StepCountter but still, there is 1 item missing in the counter :(

What exactly do you mean by that? I'd like to help you solve it.

vlastavesely commented:
I mean that console output looks like this:
```
 21/21 [============================] 100% 2 secs/2 secs 8.0 MiB - Finished!
 22/22 [============================] 100% 2 secs/2 secs 8.0 MiB - Finished!
```
Works well but on end it prints one more line. We already know where to fix it (see #905) but what file generated is doing this....? I do not know.

vlastavesely commented:
* `class-` prefix changed to `exception-`
* fixed StringRouter
* covered by tests

TomasVotruba commented:
Thank you for changes, I'll look at progress bar personally.

vlastavesely commented:
fixed cs

TomasVotruba commented:
:clap:  

Thank you, merged!

vlastavesely commented:
Have you checked the progressbar issue? It seems to be namespace-related. I just added a class from another namespace (`Nette\Loaders`) and there appeared one new line so now they are three and not just two like yesterday.
```
 26/26 [============================] 100% 8 secs/8 secs 8.0 MiB - Finished!
 27/27 [============================] 100% 8 secs/8 secs 8.0 MiB - Finished!
 28/28 [============================] 100% 8 secs/8 secs 8.0 MiB - Finished!
```
Are the sub-namespaces counted?

TomasVotruba commented:
Not yet, I'm busy with my work now so I don't have time.

I still think there will be some simple condition issue.

vlastavesely commented:
Found it! Tomorrow, I will fix it. When namespace is empty (does not contain classes), it is not counted.
